### ISSUE TITLE
BUG: random_state in shap.utils.sample not preserved between v0.41.0 -> v0.42.1

### ISSUE THREAD

**Issue by detrin** [shap/shap#3215]
### Issue Description

The result of `shap.utils.sample` function is changing even when the default random_state in https://shap.readthedocs.io/en/latest/generated/shap.utils.sample.html. I tried to find the cause of it on shap side, but didn't have much time to dig deeper for now.

I might be interested in making PR. However I don't have much time following days, I will see however what I can do.

### Minimal Reproducible Example

```python
import pandas as pd
from shap.utils import sample

X = pd.DataFrame([{'col_3': 1}, {'col_3': 0}, {'col_3': 0}, {'col_3': 0}])
masker = sample(X, 1, random_state=0)
masker.to_dict(orient="records")

# v0.41.0
# [{'col_3': 1}]

# v0.42.1
# [{'col_3': 0}]

# master b9598d5132c502d1c5368d4f517f77c0711e1c47
# [{'col_3': 0}]
```

### Expected Behavior

How does it affect us? In case that tests in other packages rely on shap values calculated on a small dataset it is an issue. In case of great size of mask created by `sample` it doesn't have a marginal effect marginally speaking. However, I believe such function should not change their output with minor releases unless it is necessary. 

### Bug report checklist

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [X] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [ ] I'd be interested in making a PR to fix this bug

### Installed Versions

```
# v0.41.1
appnope==0.1.3
asttokens==2.2.1
backcall==0.2.0
cloudpickle==2.2.1
comm==0.1.4
debugpy==1.6.7.post1
decorator==5.1.1
executing==1.2.0
ipykernel==6.25.1
ipython==8.14.0
jedi==0.19.0
joblib==1.3.2
jupyter_client==8.3.0
jupyter_core==5.3.1
llvmlite==0.40.1
matplotlib-inline==0.1.6
nest-asyncio==1.5.7
numba==0.57.1
numpy==1.23.5
packaging==23.1
pandas==2.0.3
parso==0.8.3
pexpect==4.8.0
pickleshare==0.7.5
platformdirs==3.10.0
prompt-toolkit==3.0.39
psutil==5.9.5
ptyprocess==0.7.0
pure-eval==0.2.2
Pygments==2.16.1
python-dateutil==2.8.2
pytz==2023.3
pyzmq==25.1.1
scikit-learn==1.3.0
scipy==1.11.2
-e git+https://github.com/shap/shap@510c4b6a85e4189573ef8290392139586c4292f5#egg=shap
six==1.16.0
slicer==0.0.7
stack-data==0.6.2
threadpoolctl==3.2.0
tornado==6.3.3
tqdm==4.66.1
traitlets==5.9.0
tzdata==2023.3
wcwidth==0.2.6
```

```
# v0.42.1
appnope==0.1.3
asttokens==2.2.1
backcall==0.2.0
cloudpickle==2.2.1
comm==0.1.4
debugpy==1.6.7.post1
decorator==5.1.1
executing==1.2.0
ipykernel==6.25.1
ipython==8.14.0
jedi==0.19.0
joblib==1.3.2
jupyter_client==8.3.0
jupyter_core==5.3.1
llvmlite==0.40.1
matplotlib-inline==0.1.6
nest-asyncio==1.5.7
numba==0.57.1
numpy==1.23.5
packaging==23.1
pandas==2.0.3
parso==0.8.3
pexpect==4.8.0
pickleshare==0.7.5
platformdirs==3.10.0
prompt-toolkit==3.0.39
psutil==5.9.5
ptyprocess==0.7.0
pure-eval==0.2.2
Pygments==2.16.1
python-dateutil==2.8.2
pytz==2023.3
pyzmq==25.1.1
scikit-learn==1.3.0
scipy==1.11.2
-e git+https://github.com/shap/shap@36556dd4ad3caac13c7967d9506da6a34ef6bacf#egg=shap
six==1.16.0
slicer==0.0.7
stack-data==0.6.2
threadpoolctl==3.2.0
tornado==6.3.3
tqdm==4.66.1
traitlets==5.9.0
tzdata==2023.3
wcwidth==0.2.6
```

**Comment by detrin** [comment]
I come from `probatus` even though I am not a developer of it. https://github.com/ing-bank/probatus/issues/225

**Comment by detrin** [comment]
Looking at https://github.com/scikit-learn/scikit-learn/issues/10237 and at https://github.com/shap/shap/blob/master/shap/utils/_general.py#L160. Maybe this is even out of the scope for shap and sklearn is responsible for this bug. I will check in sklearn.

**Comment by detrin** [comment]
`sklearn.utils.shuffle` https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/__init__.py#L632
`sklearn.utils.resample` https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/__init__.py#L482
uses `check_random_state` https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/__init__.py#L568

**Comment by detrin** [comment]
`sklearn.utils.validation.check_random_state`
https://github.com/scikit-learn/scikit-learn/blob/2a2772a87b6c772dc3b8292bcffb990ce27515a8/sklearn/utils/validation.py#L1207

```python
def check_random_state(seed):
    """Turn seed into a np.random.RandomState instance.

    Parameters
    ----------
    seed : None, int or instance of RandomState
        If seed is None, return the RandomState singleton used by np.random.
        If seed is an int, return a new RandomState instance seeded with seed.
        If seed is already a RandomState instance, return it.
        Otherwise raise ValueError.

    Returns
    -------
    :class:`numpy:numpy.random.RandomState`
        The random state object based on `seed` parameter.
    """
    if seed is None or seed is np.random:
        return np.random.mtrand._rand
    if isinstance(seed, numbers.Integral):
        return np.random.RandomState(seed)
    if isinstance(seed, np.random.RandomState):
        return seed
    raise ValueError(   
```

https://stackoverflow.com/a/8203401 so I think `np.random.RandomState(seed)` is used in our case

**Comment by detrin** [comment]
However, scikit-learn and numpy versions are the same, so it doesn't quite add up. The `RandomState` itself should be fine https://stackoverflow.com/a/67905877.

**Comment by detrin** [comment]
Also seems to me that the random generator is used only once in sklearn shuffle function https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/__init__.py#L590 in our case.

**Comment by thatlittleboy** [comment]
Thanks for the bug report. Seems like a serious problem, to break reproducibility in such a way.

Between 0.41.0 and 0.42+, I made the change to `shap.utils.sample` to use `sklearn.utils.shuffle` (shap 0.42) instead of `sklearn.utils.resample` (shap 0.41).  
PR here: https://github.com/dsgibbons/shap/pull/36 (this was the community fork back then, these changes were later merged into this main repo). The rationale for making this change was laid out in that PR.

We can reproduce the said issue without `shap` at all:

```python
import pandas as pd
import sklearn

X = pd.DataFrame([{'col_3': 1}, {'col_3': 0}, {'col_3': 0}, {'col_3': 0}])
print(sklearn.utils.resample(X, n_samples=1, random_state=0))  # returns row 0
print(sklearn.utils.shuffle(X, n_samples=1, random_state=0))  # returns row 2
print(sklearn.utils.resample(X, n_samples=1, random_state=0))  # returns row 0
```

with 

```
python 3.9.16

pandas==2.0.3
scikit-learn==1.3.0
numpy==1.24.4
```

So yes, it's an upstream issue (EDIT: or I should say, it's a result of different implementations of replace=True v.s. False in sklearn's `resample` function).

**Comment by thatlittleboy** [comment]
See the lines here:

https://github.com/scikit-learn/scikit-learn/blob/7f9bad99d6e0a3e8ddf92a7e5561245224dab102/sklearn/utils/__init__.py#L580C1-L585C46

Note that the only difference between `sklearn.utils.shuffle` and `sklearn.utils.resample` is that the former is `replace=True`.

We can boil down the problem to within sklearn itself:

```python
def issue_3215(replace: bool):
    # code taken from sklearn.utils.resample

    random_state = sklearn.utils.check_random_state(0)
    max_n_samples = 1  # sampling 1 item...
    n_samples = 4  # ...from a dataframe of 4 rows
    
    if replace:
        indices = random_state.randint(0, n_samples, size=(max_n_samples,))
    else:
        indices = np.arange(n_samples)
        random_state.shuffle(indices)
        indices = indices[:max_n_samples]

    return indices


replace = True
print(f"{replace=}, indices: {issue_3215(replace)}")
replace = False
print(f"{replace=}, indices: {issue_3215(replace)}")

# replace=True, indices: [0]
# replace=False, indices: [2]
```

**Comment by thatlittleboy** [comment]
@detrin So the only question left to answer is what do we do about this.

It is my opinion that the old `shap.utils.sample` was buggy. Since it is primarily used to generate a subset of the full data, for use as the background dataset, so we should sample without replacement, not with.

I'll propose this:

1. we specify clearly in the documentation for `shap.utils.sample` the sampling logic and its purpose. That it's meant to be used as sampling without replacement. I'll raise a PR for this.
2. we *could* expose a `replace` parameter for our `shap.utils.sample` function, but I think it's unnecessary. Feel free to raise your use case here though, if it's within shap's use-case, I'm open to adding it.
3. if you really wanted to use `replace=True` sampling, I suggest you can use `sklearn`'s sampling function rather than ours. Ours is just a simple wrapper anyway. If you come to the realization that you want sampling without replacement, then you can continue to use shap's `sample` function, you'll just need to do a one-time update of your tests etc.

**Comment by detrin** [comment]
@thatlittleboy Ah, I missed that change between 0.41.0 and 0.42+, now it make all sense.

> Since it is primarily used to generate a subset of the full data, for use as the background dataset, so we should sample without replacement, not with.

I agree.

From your option I would go with 1). For me personally, it doesn't matter that much. For `probatus`, it was question of changing few tests (there is already PR). In general, if there would be too many ppl complaining about this issue I would go even in later releases with 2). I think it will take a while since most of packages will switch to `shap==0.42+`.

If you make a PR, could you tag me? I would like to see how the changes in docs are made in shape.

Off topic, but what is your setup? Mine is vscode, notebooks in vscode, pyenv or brew installed python and venv.

**Comment by thatlittleboy** [comment]
@detrin I've opened a PR #3218 .

> Off topic, but what is your setup? Mine is vscode, notebooks in vscode, pyenv or brew installed python and venv.

I use neovim mostly, and occasionally jump to vscode if I need to run longer debugging sessions.  
`pyenv` for managing python versions, `virtualenv` for every project to isolate project dependencies.

**Comment by detrin** [comment]
@thatlittleboy Thank you I had a look at it.

Ad neovim, I was considering switching to neovim for a while now. Do you feel that you can be more productive in neovim than in vscode?

**Comment by thatlittleboy** [comment]
> Ad neovim, I was considering switching to neovim for a while now. Do you feel that you can be more productive in neovim than in vscode?

Yes, personally. But it does take quite a bit of setting up.  
This is off-topic for this issue, though, so let's not discuss this any further here.

**Comment by detrin** [comment]
@thatlittleboy Thanks!
